ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -295,10 +300,88 @@ public CompletableFuture<Void> downloadBlobAsync(BlobId blobId, OutputStream out
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.Unknown_Error));
+    }
+  }
+
   @Override
   public boolean deleteBlob(BlobId blobId, long deletionTime, short lifeVersion,
-      CloudUpdateValidator cloudUpdateValidator) throws CloudStorageException {
-    return false;
+      CloudUpdateValidator unused) throws CloudStorageException {
+    Timer.Context storageTimer = azureMetrics.blobUpdateDeleteTimeLatency.time();
+    AzureBlobLayoutStrategy.BlobLayout blobLayout = azureBlobLayoutStrategy.getDataBlobLayout(blobId);
+    String blobIdStr = blobLayout.blobFilePath;
+    Map<String, String> newMetadata = new HashMap<>();
+    newMetadata.put(CloudBlobMetadata.FIELD_DELETION_TIME, String.valueOf(deletionTime));
+    newMetadata.put(CloudBlobMetadata.FIELD_LIFE_VERSION, String.valueOf(lifeVersion));
+    BlobProperties blobProperties = getBlobProperties(blobLayout);
+    Map<String, String> cloudMetadata = blobProperties.getMetadata();
+    // lifeVersion must always be present
+    short cloudlifeVersion = Short.parseShort(cloudMetadata.get(CloudBlobMetadata.FIELD_LIFE_VERSION));
+    // This is the correct behavior. For ref, look at BlobStore::delete and ReplicaThread::applyUpdatesToBlobInLocalStore
+    if (cloudlifeVersion == lifeVersion && cloudMetadata.containsKey(CloudBlobMetadata.FIELD_DELETION_TIME)
+        && Long.parseLong(cloudMetadata.get(CloudBlobMetadata.FIELD_DELETION_TIME)) != Utils.Infinite_Time ) {
+      String error = String.format("Cannot delete id %s since it is already marked as deleted in cloud", blobIdStr);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.ID_Deleted));
+    }
+    // Don't rely on the CloudBlobStore.recentCache to do the "right" thing.
+    // This is the correct behavior. For ref, look at BlobStore::delete and ReplicaThread::applyUpdatesToBlobInLocalStore
+    // Repl layer handles it.
+    if (cloudlifeVersion > lifeVersion) {
+      String error = String.format("Cannot delete id %s since it has a higher lifeVersion than the message info: %s > %s",
+          blobIdStr, cloudlifeVersion, lifeVersion);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.Life_Version_Conflict));
+    }
+    cloudMetadata.putAll(newMetadata);
+    try {
+      logger.debug("Updating deleteTime of blob {} in Azure blob storage ", blobLayout.blobFilePath);
+      Response<Void> response = updateMetadata(blobLayout, blobProperties, cloudMetadata);
+      // Success rate is effective, success counter is ineffective because it just monotonically increases
+      azureMetrics.blobUpdateDeleteTimeSucessRate.mark();
+      logger.debug("Successfully updated deleteTime of blob {} in Azure blob storage with statusCode = {}, etag = {}",
+          blobLayout.blobFilePath, response.getStatusCode(), response.getHeaders().get(HttpHeaderName.ETAG));
+      return true;
+    } catch (Throwable t) {
```

Comment:
It might be helpful to catch BlobStorageExceptions separately for metrics, etc. For example:

```
catch (BlobStorageException e) {
      if (e.getErrorCode() == BlobErrorCode.CONDITION_NOT_MET) {
        azureMetrics.blobUpdateConflictCount.inc();
      }
      throw e;
    } 
```

Reply:
done